auth/token_authentication.py

```
@@ -98,3 +113,65 @@ def __call__(self, request: Request) -> ApiKey:
+    _verify_run_state(run_id, user)
+
+    return ApiKey(created_by=user, workspace=workspace)
+
+
+def _verify_run_state(run_id: str, user: AppUser) -> None:
+    from daras_ai_v2.base import BasePage, RecipeRunState
+
+    try:
+        saved_run = SavedRun.objects.get(run_id=run_id, uid=user.uid)
+    except SavedRun.DoesNotExist:
+        raise AuthenticationError("Invalid API Key. Run not found.")
```

Ruff (0.14.8)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)